packages/push/src/evaluation/cases.rs

```diff
@@ -0,0 +1,99 @@
+#[derive(Debug)]
+pub struct Case<Input, Output = Input> {
+    pub input: Input,
+    pub output: Output,
+}
+
+#[derive(Debug)]
+pub struct Cases<Input, Output = Input> {
+    cases: Vec<Case<Input, Output>>,
+}
+
+impl<Input, Output> Cases<Input, Output> {
+    #[must_use]
+    pub const fn new() -> Self {
+        Self { cases: Vec::new() }
+    }
+
+    pub fn from_inputs(
+        inputs: impl Iterator<Item = Input>,
+        target_function: impl Fn(&Input) -> Output,
+    ) -> Self {
+        inputs
+            .map(|input| Case {
+                output: target_function(&input),
+                input,
+            })
+            .collect()
+    }
+
+    // TODO: Add `from` that selects randomly from some cases
+    // TODO: Add `from` that gets cases from an external source
+    //    Maybe outside of this type?
+
+    pub fn add_case(&mut self, input: Input, output: Output) {
```

Comment:
Should this take a `Case` instead of input & output? (or a `impl Into<Case>` then one can pass `(Input, Output)` if needed)

Reply:
I definitely like that idea. We can do that in tomorrow's stream.